### ISSUE TITLE
Check walkdir performance under windows

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#9847]
Someone on reddit reports slow startup time on winrows with HDD:

https://old.reddit.com/r/rust/comments/p0z0fa/rustanalyzer_changelog_89/h8dt110/

 I don't think we should optimize for HDD, but it makes sense to make sure that we don't do something stupid on windows, as its file system is known to be exciting. 

The main disk IO should happen in this function: 

https://github.com/matklad/rust-analyzer/blob/5d5d5182c137df293e983d3c3d48869d1c21092c/crates/vfs-notify/src/lib.rs#L175-L213

curiously, we don't do any parallelism whatsoever o_O ?

**Comment by matklad** [comment]
cc @Veykril , you use windows IIRC, you might want to take a look

**Comment by matklad** [comment]
`analysis-stats` report time to load a project, and then sticking `profile` call there might help. We might also just to optimize first, and profile later -- the optimization should be simple, just pull the actual reading into the outer loop, and `.par_iter()` it.

**Comment by eminence** [comment]
I've duplicated below a comment I made on reddit.  Unfortunately I can't give access to this particular machine, but I'm happy to run any tests or development code if that's at all helpful.

------

In my experience, rust performance (and rust-analyzer performance to some extent) is predominantly impacted by disk IO speed, more than CPU speed.

For example, the sandbox repo linked above:

On a Windows machine with a fast local SSD and 8 cores, it takes 46 seconds to do a clean `cargo check` and about 16 seconds until RA finishes the "indexing" phase.

On a Linux machine with a very slow network file system and 12 cores, it takes 229 seconds to do a clean `cargo check`, and about 202 seconds until RA finishes indexing.

**Comment by matklad** [comment]
I've now realized that reading files sequentially is pretty bad architecturally, and that we might want to fix this regardless of win performance. There are cases where readling N files sequentially is very slow, while reading them in parallel is very fast.

Consider a distributed file system, where the set of files is spread across a fleet of the machines. Sequential read means N round-trip, while a fully parallel read would be just one roundtrip.

**Comment by lnicola** [comment]
There's a jwalk crate for parallel directory iteration, but I'm not sure we can use it.

**Comment by lnicola** [comment]
I wouldn't read too much into the "Database loaded" time. I often see 2-3 seconds on Linux.

**Comment by matklad** [comment]
Yeah, that's the first part: today it often includes "run `cargo check`", which shouldn't be fast. The results above are after warm-up runs and perty stable though, so I am fairly certain that there's a measurable perf gap for rust-analyzer between windows and linux. What is not clear to me is:

* is the pert due to our io, or cargo io (I *think* database load time includes cargo stating all files in the workspace, for example)
* whether we can close the gap by using windows APIs more efficiently, or whether this is an inherent "IO is slower on windows".

**Comment by lnicola** [comment]
On a warm cache, I'd expect Windows to be closer to Linux. The (worse) problems are on a cold cache, network drive etc. I still suspect parallel file reading and directory iteration are worth it, but it's not a big problem in common cases.

**Comment by memoryruins** [comment]
Since Windows Defender often slows down tools, I checked if it made any noticeable difference here. The following are stats where rust-analyzer was already built with `--release` and `cargo check` already ran once previously.

```
rust-analyzer bb1987b45 dev
cargo 1.56.0-nightly (f559c109c 2021-08-26)
```

I then used `cargo run --release -p rust-analyzer -- -q analysis-stats .` on rust-analyzer's repo for each subsequent run.

With Windows Defender **disabled**,

```
Database loaded:     1.37s
  crates: 38, mods: 770, decls: 17149, fns: 12803
Item Collection:     18.81s
  exprs: 348004, ??ty: 442 (0%), ?ty: 223 (0%), !ty: 284
Inference:           62.87s
Total:               81.68s
```

and multiple subsequent runs of the same command report roughly the same.

With Windows Defender **enabled**, Defender uses a lot of CPU for 20+ seconds until the database is reported to be loaded.

```
Database loaded:     24.85s
  crates: 38, mods: 770, decls: 17149, fns: 12803
Item Collection:     19.03s
  exprs: 348004, ??ty: 442 (0%), ?ty: 223 (0%), !ty: 284
Inference:           62.71s
Total:               81.74s
```

Subsequent runs are better with the database loading in `1.96s`, though worse than `1.37s` when disabled. Disabling then re-enabling Defender reproduces the 20+ seconds. It was also reproducible by leaving Defender enabled, restarting this desktop, then taking stats again.

If anyone else experiences this, [exclusions](https://www.howtogeek.com/671233/how-to-add-exclusions-in-windows-defender-on-windows-10/) can be added for directories like `.cargo/bin` without disabling Defender entirely.
Similar issues have been reported to cargo/rust in the past: https://github.com/rust-lang/cargo/issues/5028 / https://github.com/rust-lang/rust/issues/36277
